File: packages/core/src/test/unit/languages/javascript/fixtures/async-chains.js
// Async/await patterns

async function fetchUser(userId) {
  await new Promise(resolve => setTimeout(resolve, 100));
  return { id: userId, name: 'User' };
}

async function fetchUserDetails(userId) {
  const user = await fetchUser(userId);
  return user;
}

async function getUsers(userIds) {
  const promises = userIds.map(uid => fetchUser(uid));
  const users = await Promise.all(promises);
  return users;
}

async function processUser(userId) {
  const user = await fetchUser(userId);
  const details = await fetchUserDetails(userId);
  return { ...user, ...details };
}

class AsyncUserService {
  constructor() {
    this.cache = {};
  }

  async getUser(userId) {
    if (this.cache[userId]) {
      return this.cache[userId];
    }
    
    const user = await fetchUser(userId);
    this.cache[userId] = user;
    return user;
  }

  async updateUser(userId, data) {
    const user = await this.getUser(userId);
    Object.assign(user, data);
    return user;
  }
}

module.exports = { fetchUser, fetchUserDetails, getUsers, processUser, AsyncUserService };

File: packages/analysis/src/test/unit/languages/python/fixtures/async_chains.py
# Async/await patterns

import asyncio


async def fetch_user(user_id: str):
    """Async fetch user"""
    await asyncio.sleep(0.1)  # Simulate network delay
    return {'id': user_id, 'name': 'User'}


async def fetch_user_details(user_id: str):
    """Async fetch with await call to another async function"""
    user = await fetch_user(user_id)  # Async call
    return user


async def get_users(user_ids: list):
    """Async function calling multiple async functions"""
    tasks = [fetch_user(uid) for uid in user_ids]
    users = await asyncio.gather(*tasks)
    return users


async def process_user(user_id: str):
    """Async processing chain"""
    user = await fetch_user(user_id)
    details = await fetch_user_details(user_id)
    return {**user, **details}


class AsyncUserService:
    """Service with async methods"""
    
    def __init__(self):
        self.cache = {}
    
    async def get_user(self, user_id: str):
        """Async get user"""
        if user_id in self.cache:
            return self.cache[user_id]
        
        user = await fetch_user(user_id)
        self.cache[user_id] = user
        return user
    
    async def update_user(self, user_id: str, data: dict):
        """Async update"""
        user = await self.get_user(user_id)
        user.update(data)
        return user
